src/webhook/policy/impl/manager.go

```diff
@@ -0,0 +1,251 @@
+
+	ply, err := convertToPersistModel(policy)
+	if err != nil {
+		return 0, err
+	}
+	return dao.AddWebhookPolicy(ply)
+}
+
+// List the webhook policies, returns the total count, policy list and error
+func (m *DefaultManager) List(projectID int64) (int64, []*model.WebhookPolicy, error) {
+	var policies []*model.WebhookPolicy
+	total, persisPolicies, err := dao.GetWebhookPolicies(projectID)
+	if err != nil {
+		return total, nil, err
+	}
+
+	for _, policy := range persisPolicies {
+		ply, err := convertFromPersistModel(policy)
+		if err != nil {
+			return 0, nil, err
+		}
+		policies = append(policies, ply)
+	}
+
+	if policies == nil {
+		policies = []*model.WebhookPolicy{}
+	}
+	return total, policies, nil
+}
+
+// Get webhook policy with specified ID
+func (m *DefaultManager) Get(id int64) (*model.WebhookPolicy, error) {
+	policy, err := dao.GetWebhookPolicy(id)
+	if err != nil {
+		return nil, err
+	}
+	return convertFromPersistModel(policy)
+}
+
+// GetByNameAndProjectID webhook policy by the name and projectID
+func (m *DefaultManager) GetByNameAndProjectID(name string, projectID int64) (*model.WebhookPolicy, error) {
+	policy, err := dao.GetWebhookPolicyByName(name, projectID)
+	if err != nil {
+		return nil, err
+	}
+	return convertFromPersistModel(policy)
+}
+
+// Update the specified webhook policy
+func (m *DefaultManager) Update(policy *model.WebhookPolicy) error {
+	policy.UpdateTime = time.Now()
+	ply, err := convertToPersistModel(policy)
+	if err != nil {
+		return err
+	}
+	return dao.UpdateWebhookPolicy(ply)
+}
+
+// Delete the specified webhook policy
+func (m *DefaultManager) Delete(policyID int64) error {
+	return dao.DeleteWebhookPolicy(policyID)
+}
+
+// Test the specified webhook policy
+func (m *DefaultManager) Test(policy *model.WebhookPolicy) error {
+	p, err := json.Marshal(testPayload)
+	if err != nil {
+		return err
+	}
+
+	for _, target := range policy.Targets {
+		switch target.Type {
+		case "http":
+			return m.policyHTTPTest(target.Address, target.Secret, p)
+		default:
+			return fmt.Errorf("invalid policy target type: %s", target.Type)
+		}
+	}
+	return nil
+}
+
+func (m *DefaultManager) policyHTTPTest(address, secret string, payload []byte) error {
+	p := bytes.NewReader(payload)
+	req, err := http.NewRequest(http.MethodPost, address, p)
+	if err != nil {
+		return err
+	}
+
+	if secret != "" {
+		req.Header.Set("Authorization", "Secret "+secret)
+	}
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode > 300 {
+		return fmt.Errorf("policy test failed with response code %d", resp.StatusCode)
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Error("read policy test response body failed: %v", err)
+	}
+	log.Debugf("policy test response code %d, body: %s", resp.StatusCode, string(body))
+
+	return nil
+}
+
+func convertToPersistModel(policy *model.WebhookPolicy) (*persistModels.WebhookPolicy, error) {
```

Comment:
No any way to unify the DAO and JSON models?

Reply:
done